File: DevSupportAndroid/src/main/java/com/sup/dev/android/views/adapters/pager/PagerCardAdapter.java
package com.sup.dev.android.views.adapters.pager;

import android.content.Context;
import android.support.v4.view.PagerAdapter;
import android.view.View;
import android.view.ViewGroup;
import android.widget.FrameLayout;

import java.util.ArrayList;
import java.util.List;

import com.sup.dev.android.tools.ToolsView;
import com.sup.dev.android.views.adapters.CardAdapter;
import com.sup.dev.android.views.adapters.NotifyItem;
import com.sup.dev.android.views.cards.Card;
import com.sup.dev.java.classes.collections.HashList;

public abstract class PagerCardAdapter extends PagerAdapter implements CardAdapter{

    private final ArrayList<Holder> holders = new ArrayList<>();
    private final ArrayList<Card> items = new ArrayList<>();
    private final HashList<Class<? extends Card>, View> viewCash = new HashList<>();

    private int notifyCount = 0;

    public PagerCardAdapter() {

    }

    //
    //  items
    //

    @Override
    public Object instantiateItem(ViewGroup parent, int position) {
        Holder holder = getFreeHolder(parent);
        parent.addView(holder.itemView);

        for(int i = position; i < position + notifyCount && i < items.size(); i++)
            if(items.get(i) instanceof NotifyItem)
                ((NotifyItem)items.get(i)).notifyItem();

        Card card = items.get(position);
        FrameLayout frame = holder.itemView;

        Class<? extends Card> tag = (Class<? extends Card>) frame.getTag();

        if (frame.getChildCount() != 0)
            viewCash.add(tag, frame.getChildAt(0));
        frame.removeAllViews();

        View cardView = viewCash.removeOne(card.getClass());
        if (cardView == null)
            cardView = card.instanceView(frame.getContext());


        frame.addView(ToolsView.removeFromParent(cardView));
        frame.setTag(card.getClass());

        card.bindView(cardView);

        return holder;
    }

    @Override
    public boolean isViewFromObject(View view, Object object) {
        return object instanceof PagerCardAdapter.Holder && ((Holder) object).itemView == view;
    }

    @Override
    public void destroyItem(ViewGroup parent, int position, Object object) {
        parent.removeView(((Holder) object).itemView);
    }

    @Override
    public int getItemPosition(Object object) {
        return POSITION_NONE;
    }

    public int getItemId(int position) {
        return position;
    }

    public ArrayList<Holder> getHolders() {

        ArrayList<Holder> list = new ArrayList<>();
        for (Holder h : holders)
            list.add(h);

        return list;
    }

    //
    //  Items
    //

    @Override
    public int getCount() {
        return items.size();
    }

    public Card get(int position) {
        return items.get(realPosition(position));
    }

    public void add(Card card) {
        items.add(card);
        notifyDataSetChanged();
    }

    @Override
    public void add(int i, Card card) {
        items.add(i, card);
        notifyDataSetChanged();
    }

    public void set(List<Card> items) {
        this.items.clear();
        this.items.addAll(items);
        notifyDataSetChanged();
    }

    public void remove(Card item) {
        if (items.remove(item))
            notifyDataSetChanged();
    }

    public void remove(int position) {
        items.remove(realPosition(position));
        notifyDataSetChanged();
    }

    public void clear() {
        items.clear();
    }

    public boolean isEmpty() {
        return items.isEmpty();
    }

    public int indexOf(Card item) {
        return items.indexOf(item);
    }

    @Override
    public boolean contains(Card card) {
        return indexOf(card) > -1;
    }

    protected int realPosition(int position) {
        return position;
    }

    public int size() {
        return items.size();
    }

    public ArrayList<Card> getItems() {
        return items;
    }

    @Override
    public View getView(Card card) {
        ArrayList<Holder> holders = getHolders();
        for (Holder h : holders) if (h.item == card) return h.itemView;
        return null;
    }

    public <K extends PagerCardAdapter>K setNotifyCount(int notifyCount) {
        this.notifyCount = notifyCount;
        return (K) this;
    }

    private void removeItemFromHolders(Card item){
        for (Holder h : holders)
            if (h.item == item)
                h.item = null;
    }

    //
    //  Getters
    //

    public ArrayList<View> getViews() {

        ArrayList<View> list = new ArrayList<>();
        for (Holder h : holders)
            list.add(h.itemView);

        return list;
    }

    public View getViewIfVisible(int position) {
        for (Holder h : holders)
            if (h.item == items.get(position))
                return h.itemView;
        return null;
    }

    //
    //  Holder
    //

    private Holder getFreeHolder(ViewGroup parent) {

        for (Holder holder : holders)
            if (holder.itemView.getParent() == null)
                return holder;

        Holder holder = new Holder(parent.getContext());
        holders.add(holder);

        return holder;
    }

    protected static class Holder {

        public Card item;
        public FrameLayout itemView;

        private Holder(Context context) {
            this.itemView = new FrameLayout(context);
        }
    }
}

File: src/main/java/com/sup/dev/android/views/support/adapters/pager/PagerCardAdapter.kt
package com.sup.dev.android.views.support.adapters.pager

import android.content.Context
import androidx.viewpager.widget.PagerAdapter
import android.view.Gravity
import android.view.View
import android.view.ViewGroup
import android.widget.FrameLayout
import com.sup.dev.android.views.support.adapters.CardAdapter
import com.sup.dev.android.views.support.adapters.NotifyItem
import com.sup.dev.android.views.cards.Card
import com.sup.dev.java.classes.collections.HashList
import java.util.ArrayList

open class PagerCardAdapter : PagerAdapter(), CardAdapter {

    private val holders = ArrayList<Holder>()
    val items: ArrayList<Card> = ArrayList()
    private val viewCash = HashList<Any, View>()

    private var notifyCount = 0

    val isEmpty: Boolean
        get() = items.isEmpty()

    val views: ArrayList<View>
        get() {

            val list = ArrayList<View>()
            for (h in holders)
                list.add(h.itemView)

            return list
        }

    //
    //  Adapter
    //

    override fun notifyUpdate() {
        notifyDataSetChanged()
    }

    override fun instantiateItem(parent: ViewGroup, p: Int): Any {
        val holder = getFreeHolder(parent)
        parent.addView(holder.itemView)

        var i = p
        while (i < p + notifyCount && i < items.size) {
            if (items[realPosition(p)] is NotifyItem)
                (items[realPosition(p)] as NotifyItem).notifyItem()
            i++
        }

        val card = items[realPosition(p)]
        val frame = holder.itemView
        holder.item = card
        holder.position = p

        var cardView = viewCash.removeOne(card::class)
        if (cardView == null) cardView = card.instanceView(frame)

        frame.removeAllViews()
        frame.addView(cardView)
        frame.tag = card::class
        (cardView.layoutParams as FrameLayout.LayoutParams).gravity = Gravity.CENTER

        card.bindCardView(cardView)
        if (frame.width == 0) frame.requestLayout()

        return card
    }

    override fun isViewFromObject(view: View, obj: Any): Boolean {
        for(h in getHoldersForItem(obj)) if (h.itemView == view) return true
        return false
    }

    override fun destroyItem(parent: ViewGroup, position: Int, ob: Any) {
        for(h in getHoldersForPosition(position)){
            parent.removeView((h.itemView))
            if (h.itemView.childCount != 0) viewCash.add(h.itemView.tag, h.itemView.getChildAt(0))
            h.itemView.removeAllViews()
            h.position = -1
            h.item?.detachView()
            h.item = null
        }
    }

    override fun getItemPosition(ob: Any): Int {
        val position = indexOf(ob as Card)
        val positionX = if (position < 0) POSITION_NONE else position
        return positionX
    }

    override fun getCount(): Int {
        return items.size
    }

    //
    //  Items
    //

    fun updateAll() {
        for (c in items) c.update()
    }

    override operator fun get(i: Int): Card {
        return items[realPosition(i)]
    }

    fun add(card: Card) {
        card.setCardAdapter(this)
        items.add(card)
        notifyDataSetChanged()
    }

    override fun add(i: Int, card: Card) {
        card.setCardAdapter(this)
        items.add(i, card)
        notifyDataSetChanged()
    }


    fun set(items: List<Card>) {
        this.items.clear()
        this.items.addAll(items)
        notifyDataSetChanged()
    }

    override fun remove(card: Card) {
        if (items.remove(card)) notifyDataSetChanged()
    }

    fun remove(position: Int) {
        items.removeAt(realPosition(position))
        notifyDataSetChanged()
    }

    fun clear() {
        items.clear()
    }

    override fun indexOf(card: Card): Int {
        return items.indexOf(card)
    }

    override fun indexOf(checker: (Card) -> Boolean): Int {
        for (c in items) if (checker.invoke(c)) return indexOf(c)
        return -1
    }

    override fun <K : Card> find(checker: (Card) -> Boolean): K? {
        for (c in items) if (checker.invoke(c)) return c as K
        return null
    }

    override operator fun contains(card: Card): Boolean {
        return indexOf(card) > -1
    }

    open fun realPosition(position: Int): Int {
        return position
    }

    override fun size(): Int {
        return items.size
    }

    override fun getView(card: Card): View? {
        for (h in holders) if (h.item === card) return h.itemView.getChildAt(0)
        return null
    }

    @Suppress("UNCHECKED_CAST")
    fun <K : PagerCardAdapter> setNotifyCount(notifyCount: Int): K {
        this.notifyCount = notifyCount
        return this as K
    }

    override fun isVisible(card: Card): Boolean {
        return getView(card) != null
    }

    fun getViewIfVisible(position: Int): View? {
        for (h in holders)
            if (h.item === items[position])
                return h.itemView
        return null
    }

    //
    //  Holder
    //

    fun getHolderForView(view:View):Holder?{
        for (holder in holders) if (holder.itemView.getChildAt(0) == view) return holder
        return null
    }

    fun getHoldersForItem(item: Any): ArrayList<Holder> {
        val list = ArrayList<Holder>()
        for (holder in holders) if (holder.item == item) list.add(holder)
        return list
    }

    private fun getHoldersForPosition(position: Int): ArrayList<Holder> {
        val list = ArrayList<Holder>()
        for (holder in holders) if (holder.position == position) list.add(holder)
        return list
    }

    private fun getFreeHolder(parent: ViewGroup): Holder {
        for (holder in holders) if (holder.itemView.parent == null) return holder
        val holder = Holder(parent.context)
        holders.add(holder)

        return holder
    }

    class Holder constructor(context: Context) {

        var item: Card? = null
        var position = -1
        var itemView: FrameLayout = FrameLayout(context)

    }
}